### ISSUE TITLE
Unexpected GPU-memory behavior during SFT

### ISSUE THREAD

**Issue by JhaceLam** [modelscope/ms-swift#5230]
I’m fine-tuning Qwen3-Embedding with Swift SFT (Deepspeed + FlashAttention) and seeing three odd patterns in GPU memory:

1. **Why does GPU memory usage accumulate in discrete, phased stair-step increments?**

With per_device_train_batch_size=16, per_device_eval_batch_size=16, gradient_accumulation_steps=8, memory climbs in discrete jumps instead of staying flat (see Figure 1). What’s driving this gradual increase?

Figure 1:
<img width="667" height="451" alt="Image" src="https://github.com/user-attachments/assets/a7e39a4d-5dfb-40dd-865d-8229607f8dca" />

2. **Sudden OOM on a “similar” dataset**

I swapped in a different sample set (same number of examples and total file size) and immediately hit CUDA OOM on an H100 80 GiB—yet Figure 1 shows 80 GiB should suffice.

3. **Superlinear scaling with batch size**

When I halve per_device_train_batch_size from 16 to 8 (and double gradient_accumulation_steps from 8 to 16), peak GPU memory usage falls from ~40 GiB to ~10 GiB (Figure 2). Naively, if batch size double, then activations double, optimizer state and gradients stay constant. So why does total memory ×4?

Figure2:

<img width="658" height="451" alt="Image" src="https://github.com/user-attachments/assets/19c3ca78-34a6-4347-b638-3d1c5cd52555" />

---

Reproduction command
```bash
swift sft \
  --model Qwen3-Embedding \
  --task_type embedding \
  --model_type qwen3_emb \
  --train_type full \
  --dataset query_response_pairs.jonsl \
  --split_dataset_ratio 0.05 \
  --eval_strategy steps \
  --output_dir output \
  --eval_steps 100 \
  --num_train_epochs 1 \
  --save_steps 100 \
  --save_total_limit 5 \
  --load_best_model_at_end \
  --per_device_train_batch_size 16 \
  --per_device_eval_batch_size 16 \
  --gradient_accumulation_steps 8 \
  --learning_rate 6e-6 \
  --loss_type infonce \
  --label_names labels \
  --dataloader_drop_last true \
  --deepspeed zero2 \
  --attn_impl flash_attn
```

**Comment by 0russwest0** [comment]
**Regarding Question 1 and 2:**
Flash Attention dynamically allocates memory blocks when processing long sequences. This can lead to non-contiguous increases in memory usage. Even with similar datasets, differences in the distribution of sequence lengths can potentially cause Out of Memory (OOM) errors.

**Regarding Question 3:**
The memory usage of InfoNCE loss is proportional to the square of the batch size. This is because it requires the computation of a similarity matrix for all sample pairs within the batch.
